x/pubsub/keeper/msg_server.go

```diff
@@ -0,0 +1,360 @@
+package keeper
+
+import (
+	"context"
+	"fmt"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	stakingtypes "github.com/cosmos/cosmos-sdk/x/staking/types"
+	"github.com/peggyjv/sommelier/v4/x/pubsub/types"
+)
+
+var _ types.MsgServer = Keeper{}
+
+func (k Keeper) AddPublisherIntent(c context.Context, msg *types.MsgAddPublisherIntent) (*types.MsgAddPublisherIntentResponse, error) {
+	ctx := sdk.UnwrapSDKContext(c)
+	publisherIntent := msg.PublisherIntent
+
+	if err := publisherIntent.ValidateBasic(); err != nil {
+		return nil, sdkerrors.Wrapf(types.ErrInvalid, "invalid publisher intent: %s", err.Error())
+	}
+
+	publisher, found := k.GetPublisher(ctx, publisherIntent.PublisherDomain)
+	if !found {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrNotFound, "no publisher found with domain: %s", publisherIntent.PublisherDomain)
+	}
+
+	signer := msg.MustGetSigner()
+	if signer.String() != publisher.Address {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrUnauthorized, "registered publisher address must be signer: %s", publisher.Address)
+	}
+
+	_, found = k.GetPublisherIntent(ctx, publisherIntent.SubscriptionId, publisherIntent.PublisherDomain)
+	if found {
+		// TODO(bolten): would this UX be better if it removed the old one for you? deleting a publisher intent will clear out
+		// any subscriber intents attached to it
```

Comment:
I mean, maybe -- I feel like `AddSubscriberIntent` is effectively writing that subscriber's publisher choice for a given subscription ID, and writing it again changes it rather than manually removing then adding it. If the language is confusing, maybe `Update` makes more sense, even if it didn't originally exist? I dunno, that's why I'm asking the question.

Reply:
Currently it's written to delete the old ones if the publisher intent no longer exists, so that problem is still around in the manual deletion case. I think if we had a method of updating it while keeping the current subscriptions intact that might be a better solution.

Reply:
After thinking on it, I think the semantics of a publisher intent changing invalidates the assumptions a subscriber might have (and in fact if the ACLs change the subscriber list may no longer be valid) so we should enforce the deletion here.